### PR TITLE
Refactoring of code that uses ICoreShell.DispatchOnMainThreadAsync

### PR DESCRIPTION
**Pull request by AlexanderSher**
- Added MainThreadAwaitable and SwitchToMainThreadAsync
- Removed ICoreShell.DispatchOnMainThreadAsync because of its unclear semantics
- Fixed various threading issues (partial merge of c14dba1)

### CODE REVIEW

File: src/Common/Wpf/Impl/Threading/MainThreadAwaitable.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Threading;
+using System.Windows.Threading;
+
+namespace Microsoft.Common.Wpf.Threading {
+    public struct MainThreadAwaitable {
+        private readonly Dispatcher _dispatcher;
+
+        public MainThreadAwaitable(Thread mainThread) {
```
Comment (MikhailArkhipov) [inline src/Common/Wpf/Impl/Threading/MainThreadAwaitable.cs]: Check Thread.IsBackground ? Although it prob fails in tests

File: src/Package/Impl/DataInspect/DataImport/ImportDataWindow.xaml.cs [hunk 2]
```diff
@@ -301,7 +294,7 @@ class DataFramePreviewRowItem {
                         sw.WriteLine(line);
 
                         if (reportProgress) {
-                            var newProgressValue = 90 * (double)read / (double)sr.BaseStream.Length;
+                            var newProgressValue = 90 * (double)read / sr.BaseStream.Length;
```
Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/DataImport/ImportDataWindow.xaml.cs]: In C I've seen this first divided and then casted. C# may be different though.

File: src/Package/Impl/DataInspect/Office/CsvAppFileIO.cs [hunk 3]
```diff
@@ -46,36 +47,36 @@ internal static class CsvAppFileIO {
             var file = ProjectUtilities.GetUniqueFileName(folder, csvFileName, "csv", appendUnderscore: true);
             var rfile = file.Replace('\\', '/');
 
-            string currentStatusText = string.Empty;
-            await VsAppShell.Current.DispatchOnMainThreadAsync(() => {
-                var statusBar = VsAppShell.Current.GetGlobalService<IVsStatusbar>(typeof(SVsStatusbar));
-                statusBar.GetText(out currentStatusText);
-            });
+            string currentStatusText;
+            var statusBar = VsAppShell.Current.GetGlobalService<IVsStatusbar>(typeof(SVsStatusbar));
```
Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/Office/CsvAppFileIO.cs]: We should add Assert.IsOnMainThread to GetGlobalService

File: src/Common/Wpf/Impl/Threading/MainThreadAwaitable.cs (hunk 1 above)
Comment (AlexanderSher) [inline src/Common/Wpf/Impl/Threading/MainThreadAwaitable.cs]: Calling thread can be both UI and background (which is covered in tests)

File: src/Package/Impl/ProjectSystem/Commands/CopyItemPathCommand.cs [hunk 4]
```diff
@@ -36,19 +37,20 @@ internal sealed class CopyItemPathCommand : IAsyncCommandGroupHandler {
...
-                var directory = await _interactiveWorkflowProvider.GetOrCreate().RSession.MakeRelativeToRUserDirectoryAsync(path);
-                if (!string.IsNullOrEmpty(directory)) {
-                    await VsAppShell.Current.DispatchOnMainThreadAsync(() => {
-                        try {
-                            Clipboard.SetData(DataFormats.UnicodeText, Invariant($"\"{directory}\""));
-                        } catch (ExternalException) { }
-                    });
-                }
-                return true;
+            VsAppShell.Current.AssertIsOnMainThread();
```
Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/Commands/CopyItemPathCommand.cs]: I thought CPS does call on bg threads?

File: src/Package/Impl/DataInspect/DataImport/ImportDataWindow.xaml.cs (hunk 2 above)
Comment (AlexanderSher) [inline src/Package/Impl/DataInspect/DataImport/ImportDataWindow.xaml.cs]: AFAIK in C# unary operators have higher precedence than binary: https://msdn.microsoft.com/en-us/library/6a71f45d.aspx

File: src/Package/Impl/DataInspect/Office/CsvAppFileIO.cs (hunk 3 above)
Comment (AlexanderSher) [inline src/Package/Impl/DataInspect/Office/CsvAppFileIO.cs]: Do you want it for both unit-test and VS environments?

File: src/Package/Impl/ProjectSystem/Commands/CopyItemPathCommand.cs (hunk 4 above)
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/Commands/CopyItemPathCommand.cs]: Yeah, me too. But `InternalExecCommandAsync` which calls all the handlers is called from either `IOleCommandTarget.Exec` or `IVsUIHierarchy.ExecCommand`. AFAIK, both are called in from UI thread

File: src/Package/Impl/DataInspect/Office/CsvAppFileIO.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/Office/CsvAppFileIO.cs]: Just in the actual implementation since that IDE call hangs in bg threads.

**Comment by MikhailArkhipov** [comment]
👍

**Pull request merged**